### ISSUE TITLE
IDISP004 should warn when factory method creates an IDisposable even if the factory is an interface

### ISSUE THREAD

**Issue by CeesKaas** [DotNetAnalyzers/IDisposableAnalyzers#322]
I was evaluating IDisposableAnalyzers to see if it would be able to find a bug we had in our code but it did not so I started messing around and created the following tests where I would like to have them both pass but only the class one does.

```csharp
namespace IDisposableAnalyzers.NetCoreTests.IDISP004DontIgnoreReturnValueOfTypeIDisposableTests
{
    using System;
    using Gu.Roslyn.Asserts;

    using NUnit.Framework;

    public class ExpectWarning
    {
        private const string BaseCode = @"
namespace N
{
    using System;
    public interface IDbConnection:IDisposable{}
    public class Connection: IDbConnection{public void Dispose(){}}
InterfaceOrImplementationHere

    public static class CheckIDisposableFromFactory
    {
        private static readonly IDbConnectionFactory _dbConnectionFactory = null!;
        public static void M1()
        {
            CallTarget.TestMethod(_dbConnectionFactory.CreateConnection());
        }
    }
    public static class CallTarget
    {
        public static void TestMethod(IDbConnection connection)
        {
        }
    }
}";
        [Test]
        public void IDisposableReturnedFromImplementedFactory()
        {
            const string classDefinition = @"
    public class IDbConnectionFactory
    {
        public IDbConnection CreateConnection()
        {
            return new Connection();
        }
    }";
            var code = BaseCode.Replace("InterfaceOrImplementationHere", classDefinition, StringComparison.OrdinalIgnoreCase);

            RoslynAssert.Diagnostics(new CreationAnalyzer(), ExpectedDiagnostic.Create(Descriptors.IDISP004DoNotIgnoreCreated), code);
        }

        [Test]
        public void IDisposableFromFactoryInterface()
        {
            const string interfaceDefinition = @"
    public interface IDbConnectionFactory
    {
        IDbConnection CreateConnection();
    }";
            var code = BaseCode.Replace("InterfaceOrImplementationHere", interfaceDefinition, StringComparison.OrdinalIgnoreCase);

            RoslynAssert.Diagnostics(new CreationAnalyzer(), ExpectedDiagnostic.Create(Descriptors.IDISP004DoNotIgnoreCreated), code);
        }
    }
}
```

Our original code uses `GetConnection` but that behaves the same way
